File: apps/workbench/src/server/routers/images.ts
/**
 * Images tRPC Router
 * Handles OCR processing for images and PDFs
 */

import { z } from 'zod';
import { router, protectedProcedure } from '../trpc';
import { TRPCError } from '@trpc/server';
import { OCRService } from '../services/image/OCRService';
import { PdfService } from '../services/document/PdfService';

// Constants
const MAX_IMAGE_SIZE = 20 * 1024 * 1024; // 20MB
const MAX_PDF_SIZE = 50 * 1024 * 1024; // 50MB

/**
 * Validate buffer size and throw if exceeds limit
 */
function validateBufferSize(buffer: Buffer, maxSize: number, dataType: string): void {
  if (buffer.length > maxSize) {
    throw new TRPCError({
      code: 'BAD_REQUEST',
      message: `${dataType} too large. Maximum size: ${maxSize / (1024 * 1024)}MB, received: ${(buffer.length / (1024 * 1024)).toFixed(2)}MB`,
    });
  }
}

// Initialize services
const ocrService = new OCRService({
  provider: 'openai-vision',
  model: 'gpt-4o-mini',
});

const pdfService = new PdfService(ocrService);

export const imagesRouter = router({
  /**
   * Analyze image content with AI vision
   * Returns description and extracted information
   */
  analyzeImage: protectedProcedure
    .input(
      z.object({
        imageData: z.string(), // base64
        contentType: z.string(),
        prompt: z.string().optional(),
      })
    )
    .mutation(async ({ input }) => {
      if (!ocrService.isConfigured()) {
        throw new Error('OCR service not configured. Set OPENAI_API_KEY environment variable.');
      }

      const buffer = Buffer.from(input.imageData, 'base64');
      validateBufferSize(buffer, MAX_IMAGE_SIZE, 'Image');

      const result = await ocrService.analyzeImage(
        buffer,
        input.contentType,
        input.prompt
      );

      return result;
    }),

  /**
   * Extract text from image using OCR
   */
  extractTextFromImage: protectedProcedure
    .input(
      z.object({
        imageData: z.string(), // base64
        contentType: z.string(),
      })
    )
    .mutation(async ({ input }) => {
      if (!ocrService.isConfigured()) {
        throw new Error('OCR service not configured. Set OPENAI_API_KEY environment variable.');
      }

      const buffer = Buffer.from(input.imageData, 'base64');
      validateBufferSize(buffer, MAX_IMAGE_SIZE, 'Image');

      const result = await ocrService.extractText(buffer, input.contentType);

      return result;
    }),

  /**
   * Process PDF and extract text (with OCR if needed)
   */
  processPdf: protectedProcedure
    .input(
      z.object({
        pdfData: z.string(), // base64
        options: z
          .object({
            forceOCR: z.boolean().optional(),
            minTextThreshold: z.number().optional(),
          })
          .optional(),
      })
    )
    .mutation(async ({ input }) => {
      const buffer = Buffer.from(input.pdfData, 'base64');
      validateBufferSize(buffer, MAX_PDF_SIZE, 'PDF');

      const result = await pdfService.processPdf(buffer, input.options);

      return result;
    }),

  /**
   * Check if PDF needs OCR
   * Returns metadata and cost estimate
   */
  checkPdfNeedsOCR: protectedProcedure
    .input(
      z.object({
        pdfData: z.string(), // base64
        minTextThreshold: z.number().optional(),
      })
    )
    .query(async ({ input }) => {
      const buffer = Buffer.from(input.pdfData, 'base64');
      validateBufferSize(buffer, MAX_PDF_SIZE, 'PDF');

      const result = await pdfService.checkNeedsOCR(buffer, input.minTextThreshold);

      return result;
    }),
});

File: python/artificer_client/resources/images.py
"""
Images and OCR resource
"""

import base64
from typing import Optional
from .base import BaseResource
from ..types import OCRResult


class Images(BaseResource):
    """Images and OCR API resource."""

    def extract_text(
        self,
        image_data: bytes,
        content_type: str = "image/png"
    ) -> OCRResult:
        """
        Extract text from image using OCR.

        Uses Python OCR service for best performance.

        Args:
            image_data: Image bytes
            content_type: MIME type (e.g., "image/png", "image/jpeg")

        Returns:
            OCR result with text and confidence

        Example:
            >>> with open("receipt.jpg", "rb") as f:
            ...     result = client.images.extract_text(
            ...         f.read(),
            ...         "image/jpeg"
            ...     )
            >>> print(f"Text: {result['text']}")
            >>> print(f"Confidence: {result['confidence']}")
        """
        encoded = base64.b64encode(image_data).decode('utf-8')

        return self._trpc_request("images.extractTextFromImage", {
            "imageData": encoded,
            "contentType": content_type
        })

    def analyze(
        self,
        image_data: bytes,
        content_type: str = "image/png",
        prompt: str = "Describe this image in detail."
    ) -> dict:
        """
        Analyze image with AI vision.

        Args:
            image_data: Image bytes
            content_type: MIME type
            prompt: Analysis prompt

        Returns:
            AI analysis

        Example:
            >>> with open("chart.png", "rb") as f:
            ...     result = client.images.analyze(
            ...         f.read(),
            ...         prompt="What insights can you extract from this chart?"
            ...     )
        """
        encoded = base64.b64encode(image_data).decode('utf-8')

        return self._trpc_request("images.analyzeImage", {
            "imageData": encoded,
            "contentType": content_type,
            "prompt": prompt
        })

    def process_pdf(
        self,
        pdf_data: bytes,
        force_ocr: bool = False,
        min_text_threshold: int = 100
    ) -> dict:
        """
        Process PDF with smart OCR fallback.

        Uses Python service for 10-20x faster processing.

        Args:
            pdf_data: PDF bytes
            force_ocr: Always use OCR even if text exists
            min_text_threshold: Use OCR if extracted text < this many chars

        Returns:
            PDF text and metadata

        Example:
            >>> with open("document.pdf", "rb") as f:
            ...     result = client.images.process_pdf(f.read())
            >>> print(f"Pages: {result['result']['metadata']['pages']}")
            >>> print(f"Text: {result['result']['text'][:200]}")
        """
        encoded = base64.b64encode(pdf_data).decode('utf-8')

        return self._trpc_request("images.processPdf", {
            "pdfData": encoded,
            "options": {
                "forceOCR": force_ocr,
                "minTextThreshold": min_text_threshold
            }
        })

    def check_pdf_needs_ocr(
        self,
        pdf_data: bytes,
        min_text_threshold: int = 100
    ) -> dict:
        """
        Check if PDF needs OCR (cost estimation).

        Args:
            pdf_data: PDF bytes
            min_text_threshold: OCR threshold

        Returns:
            OCR requirement and cost estimate

        Example:
            >>> with open("scanned.pdf", "rb") as f:
            ...     check = client.images.check_pdf_needs_ocr(f.read())
            >>> if check['result']['needsOCR']:
            ...     print(f"Estimated cost: ${check['result']['estimatedCost']:.4f}")
        """
        encoded = base64.b64encode(pdf_data).decode('utf-8')

        return self._trpc_request("images.checkPdfNeedsOCR", {
            "pdfData": encoded,
            "minTextThreshold": min_text_threshold
        })
